### PR TITLE
Added more words and themes + minor bug fixes 

### PR DESCRIPTION
**Pull request by Pungent-Paws**
**PLEASE WAIT FOR MERGE OF PR Main menu and game instructions #16 INTO MASTER BEFORE MERGING THIS ONE (after review, revision, your approval)**

Minor changes to improve completeness and consistency of prints that inform user of progress in the game (Issue #18 and #19).

Adding more themes and words (Issue #2) from a txt file (which I curated simply by copying words from here http://www.english-for-students.com/Words-by-Theme.html). Words() is a class but I felt this wasn't actually necessary. Let me know what you think.

### CODE REVIEW

**Comment by arumoy-shome** [comment]
I do feel that making words a class is overkill since we don't need to make instances out of it. 
Additionally, imho, the original implementation of words.py containing a dict was nice and simple. We don't really need that many themes and words anyway (KISS). :)

**Comment by arumoy-shome** [comment]
On a separate note, let's keep one PR per issue (unless they are closely related, then on PR per theme). This makes it easy to revert the PR incase things go south without losing additional functionality that the PR may have introduced in a seemingly unrelated part of the codebase :)

**Comment by arumoy-shome** [comment]
@Pungent-Paws looks like you need to rebase this branch off of master to get rid of rohit and my commits!

**Comment by Pungent-Paws** [comment]
> 1. Issue #19 still exists :(
>    image
>    image
> 2. #18 seems fine
>    image
> 3. New themes and words (#2) work great! Creating an entire class for it probably is overkill.

@shaw-rohit  I think the 2nd pic shows the correctly guessed word (iow #19 has been resolved), no?

Other than that I agree with you and @arumoy-shome about not needing a class for the words. And in hindsight I'm also questioning the need for so many themes and words, I guess we could indeed stick to what we originally had (a small dictionary with cities & flowers and a couple of words for them each).

**Comment by shaw-rohit** [comment]
> > 1. Issue #19 still exists :(
> >    image
> >    image
> > 2. #18 seems fine
> >    image
> > 3. New themes and words (#2) work great! Creating an entire class for it probably is overkill.
> 
> @shaw-rohit I think the 2nd pic shows the correctly guessed word (iow #19 has been resolved), no?
> 
> Other than that I agree with you and @arumoy-shome about not needing a class for the words. And in hindsight I'm also questioning the need for so many themes and words, I guess we could indeed stick to what we originally had (a small dictionary with cities & flowers and a couple of words for them each).

Umm, it doesn't. It shows `B U T T E R C U _` but should be `B U T T E R C U P`.

**Comment by Pungent-Paws** [comment]
> @Pungent-Paws looks like you need to rebase this branch off of master to get rid of rohit and my commits!
@arumoy-shome  Yes, need a bit of time though, never done that before


> > > 1. Issue #19 still exists :(
> > >    image
> > >    image
> > > 2. #18 seems fine
> > >    image
> > > 3. New themes and words (#2) work great! Creating an entire class for it probably is overkill.
> > 
> > 
> > @shaw-rohit I think the 2nd pic shows the correctly guessed word (iow #19 has been resolved), no?
> > Other than that I agree with you and @arumoy-shome about not needing a class for the words. And in hindsight I'm also questioning the need for so many themes and words, I guess we could indeed stick to what we originally had (a small dictionary with cities & flowers and a couple of words for them each).
> 
> Umm, it doesn't. It shows `B U T T E R C U _` but should be `B U T T E R C U P`.

Ah yeah I see what you mean, I'll adjust

**Comment by Pungent-Paws** [comment]
It might just be sleepiness, but I am rethinking (/ overthinking) my understanding of the concept of rebasing and because I'm not sure anymore I'm scared to and unsure how to do it (for the first time).

So this is how I'm understanding (or: being confused about) the situation:
This wm-theme_and_tracking branch should eventually contain approved changes we'd like to merge into master. We don't want to 'contaminate' the master with too many or duplicate commits, so we want to ''get rid of commits'' in this branch (these commits are also in the master's commit history). 

But then if I rebase this branch off of master, this branch will have a commit history identical to the master's, no? And then later when merging this branch into master, won't we have duplicate commits?

I'm probably misunderstanding something, I'm sorry :(

**Comment by shrutirao94** [comment]
I'm assuming you'll follow the following steps:
1. pull from master
2. checkout your own local branch
3. `git rebase master` from your local branch

So Arumoy and Rohit's commits will be applied to the master in your local machine when you pull from remote master. Next, when you rebase your local branch from master on your machine, their commits will be added to your branch. Your commits (given no clash with theirs) will be added on top (based on how the events occurred chronologically). So your local branch should look exactly like master, except for your new commits! You can use `git log` to see the all the commits that were added! When you rebase, you only need concern yourself with conflicts that might occur. You don't need to worry about anything else. Git will deal with all the admin stuff you are concerned about. 

Now, after rebase, when you push your branch back to master (given no conflicts), git in the background will check for all commits (via a hash) and only apply the new/updated ones.

 So tl;dr all of that admin stuff  gets handled by git!

I hope this helps a bit?

File: hangman/interface.py [hunk 1]
```diff
@@ -71,32 +71,30 @@ def write_hint(self, hint, theme):
 
     def write_health(self, health):
         print("\nRemaining lives: {0}".format(health))
-        print("Remaining lives: {0}".format(health))
 
     def write_guessed_letters(self, guesses):
-        print("Letters guessed: {0}".format(guesses))
+        print("Letters guessed: {0}".format(guesses)) if len(guesses) > 0 else print("Letters guessed: [ ]")
```
Comment (arumoy-shome) [inline hangman/interface.py:76]: don't think you need this change anymore since `player.guesses` was changed back to a list.

File: hangman/interface.py [hunk 2]
```diff
@@ -71,32 +71,30 @@ def write_hint(self, hint, theme):
...
 
     def write_already_guessed(self):
         print("You have already guessed this letter! Try again!")
 
-    def write_incorrect_guess(self, guess):
-        print("\nSorry! {0} is not in the word! Try again!".format(guess))
+    def write_incorrect_guess(self, guess, health):
+        print("\nSorry! {0} is not in the word! Try again!".format(guess)) if health > 0 else print("\nSorry! {0} is not in the word!".format(guess))
```
Comment (arumoy-shome) [inline hangman/interface.py:82]: I don't think this change is required either because the same statement is printed for both if-else conditions?

Comment (Pungent-Paws) [reply hangman/interface.py:82]: When health becomes 0, there are no more tries to be made, so I thought it might be nicer to not print ''Try again'' as done when there are still guessed to be made. 
Untitled

Comment (arumoy-shome) [reply hangman/interface.py:82]: Ah my bad, I missed the "try again" part. Instead of passing the player health into the interface (which does not seem sound from a design stand point) how about we introduce a `try` param to `def write_incorrect_guess` which can default to 1 to indicate that the player can try again? Then we move the health logic back into game.py and manipulate `try`?

Comment (Pungent-Paws) [reply hangman/interface.py]: > Ah my bad, I missed the "try again" part. Instead of passing the player health into the interface (which does not seem sound from a design stand point) how about we introduce a `try` param to `def write_incorrect_guess` which can default to 1 to indicate that the player can try again? Then we move the health logic back into game.py and manipulate `try` (set it to 0 if player.health == 0)?

I added an attribute to Game() called `attempts_left` which is initialized as True and becomes Fault when player health becomes 0. So we don't pass player health logic to interface.py anymore, but a game attribute. But I'm not sure if I fully udnerstood the design issue you had with the previous (quoted) solution and if this new suggestion actually addresses it?

Comment (arumoy-shome) [reply hangman/interface.py]: It's getting there :). My suggestion was to decouple two seemingly unrelated classes from each other (player and interface). The interface class is meant to facilitate communication with the person playing the game, thus it should not rely on any info. from the `Game` and `Player` classes. So the new changes do not address the core issue (now instead of relying on Player, we rely on Game). My suggestion was to do something like this:

```python
def write_incorrect_guess(self, try=True):
    if try:
        # print something
    else:
        # print something else
```

And then from game we can do:

```python
if self.player.gethealth() > 0:
    self.interface.write_incorrect_guess()
else:
    self.interface.write_incorrect_guess(try=False)
```

This way `Interface` remains decoupled from the other classes. Hope this help! Let me know if I can help!

File: hangman/interface.py (hunk 1 above)
Comment (arumoy-shome) [reply hangman/interface.py:76]: Reminder that this still needs to be removed!

File: hangman/interface.py (hunk 2 above)
Comment (Pungent-Paws) [reply hangman/interface.py]: Thanks for taking the time to explain, I think I understand the difference! Updated accordingly (sorry about the long commit history though)

**Comment by Pungent-Paws** [comment]
My latest push late last night contains, if I'm not mistaken, all requested changes. I didn't merge it into master yet (but did already rebase off of master). Ready to merge into master as soon as I have approval of one of you  @arumoy-shome  @shaw-rohit

**Comment by arumoy-shome** [comment]
I disagree with @shaw-rohit ‘s. I don’t think we should duplicate those print statements inside the if clause as well. I think the current implementation where `interface.write_won()` mentions the solution is good and mirrors what `interface.write_lost()` does.

**Comment by shaw-rohit** [comment]
> I disagree with @shaw-rohit ‘s. I don’t think we should duplicate those print statements inside the if clause as well. I think the current implementation where `interface.write_won()` mentions the solution is good and mirrors what `interface.write_lost()` does.

What you're suggesting looks good from a programmer's perspective (clean code!) but UX should have higher priority since it's a game :D 

Moreover, once we have the graphical hangman in place, we wouldn't wanna end the game with a 3-limbed (handicapped) hangman, right? 

The win-win solution I can think of: Putting the print statements in a method and calling it in the while loop as well as the won/lost conditions.

**Comment by arumoy-shome** [comment]
@Pungent-Paws thanks for incorporating the changes, I think the PR looks good and ready to 🐑
